src/include/sof/audio/audio_stream.h

```diff
@@ -203,6 +215,51 @@ static inline uint32_t audio_stream_sample_bytes(const struct audio_stream __spa
 	return get_sample_bytes(buf->frame_fmt);
 }
 
+/**
+ * Return the frames that meet the align requirement of both byte_align and
+ * frame_align_req.
+ * @param byte_align Processing byte alignment requirement.
+ * @param frame_align_req Processing frames alignment requirement.
+ * @param frame_size Size of the frame in bytes.
+ * @return frame number.
+ */
+static inline uint32_t audio_stream_frame_align_get(uint32_t byte_align, uint32_t frame_align_req,
+						    uint32_t frame_size)
+{
+	/* Figure out how many frames are needed to meet the byte_align alignment requirements */
+	uint32_t frame_num = byte_align / gcd(byte_align, frame_size);
+
+	/** return the lcm of frame_num and frame_align_req*/
+	return frame_align_req * frame_num / gcd(frame_num, frame_align_req);
+}
+
+/**
+ * Set frame_align_shift and frame_align of stream according to byte_align and
+ * frame_align_req alignment requirement. Once the channel number,frame size
+ * are determined，the frame_align and frame_align_shift are determined too.
+ * these two feature will be used in audio_stream_get_avail_frames_aligned
+ * to calculate the available frames. it should be called in component prepare
+ * or param functions only once before stream copy. if someone forgets to call
+ * this first, there would be unexampled error such as nothing is copied at all.
+ * @param byte_align Processing byte alignment requirement.
+ * @param frame_align_req Processing frames alignment requirement.
+ * @param stream Sink or source stream structure which to be set.
+ */
+static inline void audio_stream_init_alignment_constants(uint32_t byte_align,
+							 uint32_t frame_align_req,
+							 struct audio_stream *stream)
+{
+	uint32_t process_size;
+	uint32_t frame_size = audio_stream_frame_bytes(stream);
+
+	stream->frame_align = audio_stream_frame_align_get(byte_align, frame_align_req, frame_size);
+	process_size = stream->frame_align * frame_size;
+	if (is_power_of_2(process_size))
+		stream->frame_align_shift = 31 - clz(process_size);
+	else
+		stream->frame_align_shift = 32 - clz(process_size);
```

Comment:
```suggestion
		stream->frame_align_shift =
		 is_power_of_2(process_size) ? 31 : 32 - clz(process_size);
```

Reply:
@marc-hb did you mean
```
	stream->frame_align_shift =
		(is_power_of_2(process_size) ? 31 : 32) - clz(process_size);
```
?

Reply:
Yes of course, thanks @lyakh for correcting my C precedence mistake.

I see @andrula-song also corrected it in 3cdc49c3f2e652cef